### ISSUE TITLE
Converting circular structure to JSON in logger handling of Exception

### ISSUE THREAD

**Issue by lspellman-opsguru** [sensedeep/dynamodb-onetable#289]
I doubt this is an issue with code, but with the construction of the model, but I've not been able to figure it out. Hate anyone having to spend time on this, but appreciate any quick pointers.

When creating a new item for table, the following stack trace is generated:
```
create item error: TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'IncomingMessage'
    |     property 'req' -> object with constructor 'ClientRequest'
    --- property 'res' closes the circle
    at JSON.stringify (<anonymous>)
    at Log.logger (file:///var/task/src/handlers/process-badge-scans.js:22:26)
    at Log.process (file:///var/task/node_modules/dynamodb-onetable/dist/mjs/Table.js:1003:18)
    at Log.error (file:///var/task/node_modules/dynamodb-onetable/dist/mjs/Table.js:993:14)
    at Table.execute (file:///var/task/node_modules/dynamodb-onetable/dist/mjs/Table.js:532:30)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Model.run (file:///var/task/node_modules/dynamodb-onetable/dist/mjs/Model.js:321:22)
    at async Model.putItem (file:///var/task/node_modules/dynamodb-onetable/dist/mjs/Model.js:768:16)
    at async Model.create (file:///var/task/node_modules/dynamodb-onetable/dist/mjs/Model.js:485:22)
    at async addRecords (file:///var/task/src/handlers/process-badge-scans.js:61:21)
```

Here is the model it is working on:
```
{
  "indexes": {
    "primary": {
      "hash": "PK",
      "sort": "SK"
    }
  },
  "models": {
    "CLASS": {
      "type": {
        "type": "string",
        "value": "CLASS"
      },
      "PK": {
        "type": "string",
        "value": "TENANT#${tenantId}#${className}"
      },
      "SK": {
        "type": "string",
        "value": "${className}"
      },
      "className": {
        "type": "string"
      },
      "description": {
        "type": "string"
      },
      "tenantId": {
        "type": "string"
      }
    },
    "SECTION": {
      "type": {
        "type": "string",
        "value": "SECTION"
      },
      "PK": {
        "type": "string",
        "value": "TENANT#${tenantId}#${className}"
      },
      "SK": {
        "type": "string",
        "value": "${className}#${sectionNumber}"
      },
      "className": {
        "type": "string"
      },
      "sectionNumber": {
        "type": "number"
      },
      "timeStart": {
        "type": "string"
      },
      "timeEnd": {
        "type": "string"
      },
      "teacher": {
        "type": "string"
      },
      "classroom": {
        "type": "string"
      },
      "weekdays": {
        "type": "Object", "schema": {
          "weekdays": { "type": "Array"}
        }
      },
      "tenantId": {
        "type": "string"
      }
    },
    "ATTENDANCE": {
      "type": {
        "type": "string",
        "value": "ATTENDANCE"
      },
      "PK": {
        "type": "string",
        "value": "TENANT#${tenantId}#${className}"
      },
      "SK": {
        "type": "string",
        "value": "${className}#${sectionNumber}#${classDate}"
      },
      "className": {
        "type": "string"
      },
      "sectionNumber": {
        "type": "number"
      },
      "studentId": {
        "type": "string"
      },
      "badgeInTime": {
        "type": "string"
      },
      "classDate": {
        "type": "string"
      },
      "tenantId": {
        "type": "string"
      }
    },
    "STUDENT": {
      "type": {
        "type": "string",
        "value": "STUDENT"
      },
      "PK": {
        "type": "string",
        "value": "TENANT#${tenantId}#STUDENT"
      },
      "SK": {
        "type": "string",
        "value": "${lastName}#${firstName}"
      },
      "studentId": {
        "type": "number"
      },
      "firstName": {
        "type": "string"
      },
      "lastName": {
        "type": "string"
      },
      "classOf": {
        "type": "number"
      },
      "tenantId": {
        "type": "string"
      }
    },
    "TEACHER": {
      "type": {
        "type": "string",
        "value": "TEACHER"
      },
      "PK": {
        "type": "string",
        "value": "TENANT#${tenantId}#TEACHER"
      },
      "SK": {
        "type": "string",
        "value": "${lastName}#${firstName}"
      },
      "employeeId": {
        "type": "string"
      },
      "firstName": {
        "type": "string"
      },
      "lastName": {
        "type": "string"
      },
      "department": {
        "type": "string"
      },
      "tenantId": {
        "type": "string"
      }
    }
  },
  "version": "0.1.0",
  "params": {
    "typeField": "type"
  },
  "format": "onetable:1.0.0",
  "queries": {}
};
```

relevant code in lambda:
```
async function addRecords(rec, model) {
    try {
        const res = await model.create(rec, { log: true });
        return res;
    } catch (err) {
        console.log("create item error:", err);
    }
}
```

**Comment by lspellman-opsguru** [comment]
More Info: removing SK from the ConditionExpression and ExpressionAttributeNames allows it to succeed:
```
ConditionExpression: '(attribute_not_exists(#_0))',
ExpressionAttributeNames: { '#_0': 'PK' },
```
Question: are there use-cases where on a put, both of those NEED to be evaluated?  Referencing Alex DeBrie, it appears to only be a potential cause of issues rather than providing value. Quoting:

**Comment by lspellman-opsguru** [comment]
In the Expression.js file, addConditions method, I commented out the two places where it added in a condition for the sort key. Tried new item, worked fine.

Interestingly though, when then sending a dupe of that item, it fails (as it should) with a Conditional check failure. AND then, it proceeds to generate a create item error call stack that references the "Converting circular structure to JSON..."  message. So, that made me think...is the error on the put operation itself...or on reporting the error of the put not being successful?  

Doing a little more testing, I reset the code back to default, made sure Dynamo table did NOT have an entry in it...submitted, and it works fine.  Submitted again (creating a dupe possibility) and, when it fails, sure enough, the circular error is in how the code is reporting the Conditional check failure -- not as the reason for the put failure.

So I think when I originally reported the error, I may have not realized I'd already added the items to the table, and when code ran -- it failed, because of the Conditional check. But in trying to report that to me, the logger ran into this error and reported it.  

### TL;DR -- Check your error handling code for how your reporting from Table.js in execute method, that appears to be where the issue is.

**Comment by mobsense** [comment]
Thanks for reporting. We'll dig into this over the next week. A little swamped for a few days, but then it will get good attention.

**Comment by mobsense** [comment]
Can you please create a minimal debug.ts that replicates the problem with the schema and code to trigger the issue. That will accelerate our investigations.

**Comment by mobsense** [comment]
Closing till you can provide a bit more info. Please feel free to reopen when you have the time.

**Comment by revmischa** [comment]
I'm getting this exact error

My schema:
```ts
import { OneSchema } from "dynamodb-onetable";

const Match = {
  ulid: /^[0-9A-Z]{26}$/i,
  email:
    /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
};

const id = {
  type: String,
  required: true,
  generate: "ulid",
  validate: Match.ulid,
} as const;

export const DbSchema = {
  format: "onetable:1.1.0",
  version: "0.0.1",

  indexes: {
    primary: { hash: "pk", sort: "sk" },
    gs1: {
      hash: "gs1pk",
      sort: "gs1sk",
      project: ["gs1pk", "gs1sk"],
      follow: true,
    },
  },

  models: {
    User: {
      pk: { type: String, value: "user:${id}" },
      sk: { type: String, value: "user-email:${email}" },
      id: { type: String, required: true },
    },

    Message: {
      pk: { type: String, value: "message:${id}" },
      sk: { type: String, value: "message-send-after:${sendAfter}" },

      id,

      userId: { type: String, required: true }, // sending user
      sendAfter: { type: Number, default: 0, required: true }, // timestamp of when to send
      messageParams: { type: String, required: true }, // JSON string of message params
      isSent: { type: Boolean, default: false, required: true }, // true if sent
      sentAt: { type: Number, required: false }, // timestamp of when sent

      gs1sk: { type: String, value: "message-is-sent:${isSent}" },
    },
  } as const,

  params: {
    isoDates: true,
    timestamps: true,
  },
} as const;

// typecheck schema here
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const validate: OneSchema = DbSchema;
```

Triggering error:
```ts
const message = await Message.create({
    userId: "ABC",
    messageParams: JSON.stringify({"to": "foo@bar.gov"}),
    isSent: false,
  });

...
  const message = await Message.get({ id: message.id });
```

**Comment by mobsense** [comment]
That works fine for me with the latest version.

Here is my debug.ts

```
/*
    debug.ts - Just for debug

    Edit your test case here and invoke via: "jest debug"

    Or run VS Code in the top level directory and just run.
 */
import {AWS, Client, Entity, Model, Table, print, dump, delay} from './utils/init'
import { OneSchema } from '../src/index.js'

jest.setTimeout(7200 * 1000)

const Match = {
    ulid: /^[0-9A-Z]{26}$/i,
    email:
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
  };
  
  const id = {
    type: String,
    required: true,
    generate: "ulid",
    validate: Match.ulid,
  } as const;
  
  const schema = {
    format: "onetable:1.1.0",
    version: "0.0.1",
  
    indexes: {
      primary: { hash: "pk", sort: "sk" },
      gs1: {
        hash: "gs1pk",
        sort: "gs1sk",
        project: ["gs1pk", "gs1sk"],
        follow: true,
      },
    },
  
    models: {
      User: {
        pk: { type: String, value: "user:${id}" },
        sk: { type: String, value: "user-email:${email}" },
        id: { type: String, required: true },
      },
  
      Message: {
        pk: { type: String, value: "message:${id}" },
        sk: { type: String, value: "message-send-after:${sendAfter}" },
  
        id,
  
        userId: { type: String, required: true }, // sending user
        sendAfter: { type: Number, default: 0, required: true }, // timestamp of when to send
        messageParams: { type: String, required: true }, // JSON string of message params
        isSent: { type: Boolean, default: false, required: true }, // true if sent
        sentAt: { type: Number, required: false }, // timestamp of when sent
  
        gs1sk: { type: String, value: "message-is-sent:${isSent}" },
      },
    } as const,
  
    params: {
      isoDates: true,
      timestamps: true,
    },
  } as const;

//  Change your table params as required
const table = new Table({
    name: 'DebugTable',
    client: Client,
    schema,
    logger: true,
})

type UserType = Entity<typeof schema.models.User>
let User = table.getModel<UserType>('User')
let user: UserType = null

type MessageType = Entity<typeof schema.models.Message>
let Message = table.getModel<MessageType>('Message')
let msg: MessageType = null


//  This will create a local table
test('Create Table', async() => {
    if (!(await table.exists())) {
        await table.createTable()
        expect(await table.exists()).toBe(true)
    }
})

test('Test', async() => {
    const validate: OneSchema = schema
    
    let message = await Message.create({
        userId: "ABC",
        messageParams: JSON.stringify({"to": "foo@bar.gov"}),
        isSent: false,
      }, {log: true});
    
      message = await Message.get({id: message.id}, {log: true});
    
})

test('Destroy Table', async() => {
    await table.deleteTable('DeleteTableForever')
    expect(await table.exists()).toBe(false)
})

```

**Comment by mobsense** [comment]
Tested with: 49784833414398a0d7f9b467a348c202085c7037 from the repo

**Comment by revmischa** [comment]
> Tested with: [4978483](https://github.com/sensedeep/dynamodb-onetable/commit/49784833414398a0d7f9b467a348c202085c7037) from the repo

Ok sorry the problem wasn't in the `get` call it was in this following `update` call:

```ts
    await Message.update({ id: message.id, isSent: true });
```

I'm able to trigger the error running your debug.ts with that line added at the end

**Comment by mobsense** [comment]
This is why our issue template requests a debug.ts

**Comment by mobsense** [comment]
That works for us here.

Can you follow that through in a debugger in the debug.ts and set a break point at 575 in Table.js.

Then see what field of the context is failing the JSON.stringify()

We should come up with a general solution, but I need to know what field has the loop.

**Comment by revmischa** [comment]
<img width="1640" alt="Screen Shot 2022-04-05 at 11 14 06" src="https://user-images.githubusercontent.com/245131/161732142-0434ad0d-d7fa-4cd4-902f-94cf3fd2dc97.png">

**Comment by mobsense** [comment]
Can you trace what field has the loop?

Do JSON.stringify() on various properties in the context to spot the culprit.

**Comment by lspellman-opsguru** [comment]
I got back into my project yesterday to update some things. Prior, I'd
pulled the logger out completely due to the issue and it obscurring
root problems rather than helping. I haven't really been paying close
attention to this thread, but will review it now to see if I can
assist with isolation.

Lance

**Comment by mobsense** [comment]
We've pushed a work around to the repo which does a try/catch around the log and then tries to emit each property individually.

HOWEVER: I'd love to know which property is causing you grief and is circular.

**Comment by travtarr** [comment]
I've recently upgraded `@aws-sdk/client-dynamodb`  to version 3.58.0 from 3.52.0 and that has fixed this issue for me.

I then went back and installed each version between 3.52.0 and 3.58.0 and found that it was resolved in [3.55.0](https://github.com/aws/aws-sdk-js-v3/releases/tag/v3.55.0). After looking through those commits I'm not too sure what changed to fix this issue.

**Comment by mobsense** [comment]
Fixed in 2.3.5
